scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -292,7 +292,7 @@ s32 BPF_STRUCT_OPS(rustland_select_cpu, struct task_struct *p, s32 prev_cpu,
 	 * dispatch directly from .enqueue(), bypassing the user-space
 	 * scheduler.
 	 */
-	tctx->force_local = get_cpu_owner(prev_cpu) == 0;
+	tctx->force_local = scx_bpf_test_and_clear_cpu_idle(prev_cpu);
```

Comment:
Isn't this still racy because the rust userland component doesn't participate? `select_cpu` may decide to put a task on `prev_cpu` according to builtin idle tracking but before that becomes visible to userland, it may also decide to assign another task to the same CPU. Getting rid of the race would require implementing atomic op across BPF and userland, no?

Reply:
Yeah, it's still racy given that user space can dispatch a task to that core. My thinking is that this makes it slightly less racy by getting rid of the race between context switching, but perhaps we should just keep it the way it is as it's inevitable that we'll need to implement some kind of idle tracking between user space and the kernel anyways?

Reply:
> Yeah, it's still racy given that user space can dispatch a task to that core. My thinking is that this makes it slightly less racy by getting rid of the race between context switching, but perhaps we should just keep it the way it is as it's inevitable that we'll need to implement some kind of idle tracking between user space and the kernel anyways?

Thinking more about this, given that both solutions are racy in a way or another, I think I'm more inclined to maintain the current setup and (try to) rely as much as possible on the cpu ownership map. Just because it easily gives the possibility to the user-space scheduler to reserve some CPUs to certain PIDs (like creating potential reservation mechanism to lock certain CPUs to certain PIDs, or implement similar ideas). We don't have to necessarily do that, but having the possibility to do that would be nice, also considering that this scheduler was designed as a foundational framework for implementing more complex scheduling policies in Rust.

Reply:
It *should* be possible to make it not racy by using atomic ops from both bpf and rust. Making data types matching might be tricky tho.

Reply:
That all makes sense. I dropped the change to scx_rustland.

Reply:
> It _should_ be possible to make it not racy by using atomic ops from both bpf and rust. Making data types matching might be tricky tho.

Yeah I think it's possible. Unfortunately we're back to the classic impedance problem of cpumask kptrs in the kernel vs. how they're represented in user space